src/blog/understanding-ecmascript-part-2.md

```diff
@@ -0,0 +1,342 @@
+
+We can also use the property access as a right hand side of an assignment:
+
+```javascript
+x = o2.foo;
+```
+
+In this case, the behavior is defined in the runtime semantics for the `AssignmentExpression : LeftHandSide = AssignmentExpression` production. Also this ends up calling `GetValue` on the result of evaluating the right hand side `AssignmentExpression`.
+
+> [Runtime Semantics: Evaluation for `AssignmentExpression : LeftHandSideExpression = AssignmentExpression`](https://tc39.es/ecma262/#sec-assignment-operators-runtime-semantics-evaluation)
+>
+> 1. If `LeftHandSideExpression` is neither an `ObjectLiteral` nor an `ArrayLiteral`, then
+>     1. Let `lref` be the result of evaluating `LeftHandSideExpression`.
+>     1. `ReturnIfAbrupt(lref)`.
+>     1. If `IsAnonymousFunctionDefinition(AssignmentExpression)` and `IsIdentifierRef` of `LeftHandSideExpression` are both `true`, then
+>         1. Let `rval` be `NamedEvaluation` of `AssignmentExpression` with argument `GetReferencedName(lref)`.
+>     1. Else,
+>         1. Let `rref` be the result of evaluating `AssignmentExpression`.
+>         1. Let `rval` be `? GetValue(rref)`.
+>     1. Perform `? PutValue(lref, rval)`.
+>     1. Return `rval`.
+> 1. Let `assignmentPattern` be the `AssignmentPattern` that is covered by `LeftHandSideExpression`.
+> 1. Let `rref` be the result of evaluating `AssignmentExpression`.
+> 1. Let `rval` be `? GetValue(rref)`.
+> 1. Perform `? DestructuringAssignmentEvaluation` of `assignmentPattern` using `rval` as the argument.
+> 1. Return `rval`.
+
+Now the `LeftHandSideExpression` (`x`) is not an object literal or an array literal, so we take the if branch in step 1. In step 1 a, we evaluate the left-hand side (`x`) and store the result into `lref`. `o2.foo` is not a function definition, so we don't take the if branch in 1 c, but the else branch in 1 d. There we evaluate `o2.foo` and call `GetValue` on it.
+
+In any case, `GetValue` will be called on the Reference which is the result of evaluating `o2.foo`. Thus, we know that the Object internal method `[[Get]]` will get invoked when accessing a property on an Object, and the prototype chain walk will occur.
+
+### Why is `o2.foo` an `AssignmentExpression`?
+
+Just one more thing. Previously, we used `o2.foo` as an `AssignmentExpression`. It doesn't look like one though. Why is it an `AssignmentExpression`?
+
+The spec actually allows an `AssignmentExpression` both as an argument and as the right hand side of an assignment. For example:
+
+```javascript
+function simple(a) { console.log('The argument was ' + a); }
+simple(x = 1); // Prints out: The argument was 1
+x; // 1
+```
+
+and
+
+```javascript
+x = (y = 5);
+x; // 5
+y; // 5
+```
+
+`o2.foo` is a "degenerate" `AssignmentExpression` which doesn't assign anything. This follows from the following grammar productions, each one taking the "simplest" or "most degenerate" case until the last one:
```

Comment:
What does degenerate mean here? The productions below and the layering encodes operator precedence and how closely expressions bind, not about simplicity.

Reply:
As discussed, removed the word "generate" and explained it differently